test/Parse/recovery.swift

```diff
@@ -455,13 +455,17 @@ struct ErrorTypeInVarDeclArrayType3 {
 
 struct ErrorTypeInVarDeclArrayType4 {
   var v1 : Int[1 // expected-error {{expected ']' in array type}} expected-note {{to match this opening '['}}
+  var v2 : Int] // expected-error {{unexpected ']' in type; did you mean to write an array type?}} {{12-12=[}}
```

Comment:
Some random complex cases I'd like to see tests of:

```swift
Swift.Int]
Set<Int]>
Set<Int>]
Int]?
Int?]
[Int]]
[String: Int]]
```

I'm guessing some of these will still have pretty bad diagnostics; if there isn't a blatantly simple fix, you can just FIXME those and take the incremental improvement.

Reply:
Sure, will do!

Reply:
Tests added:

`Set<Int]>` complains of "expected '>' to complete generic argument list" instead.
`Int]?` diagnoses the missing `[` but doesn't consume the trailing '?', so then additional errors about consecutive declarations on a line.

The rest of these examples work as you'd like.